ikarus/finiteelements/feresulttypes.hh

```diff
@@ -0,0 +1,189 @@
+    using Vectorizer = VectorizeStruct;                                                                     \
+    using Matricizer = MatricizeStruct;                                                                     \
+  }
+
+#define REGISTER_SYMMETRIC_RESULTTYPE(structName, rowsExpr, ColsExpr, MaxRowsExpr, MaxColsExpr) \
+  REGISTER_RESULTTYPE_IMPL(structName, rowsExpr, ColsExpr, Ikarus::Util::VectorizeWithVoigt,    \
+                           Ikarus::Util::MatricizeWithVoigt)
+
+#define REGISTER_SIMPLE_SYMMETRIC_RESULTTYPE(structName, rowsExpr, ColsExpr, strainlike) \
+  REGISTER_RESULTTYPE_IMPL(structName, rowsExpr, ColsExpr, rowsExpr, ColsExpr,           \
+                           Ikarus::Impl::VectorizeWithVoigt<strainlike>, Ikarus::Impl::MatricizeWithVoigt<strainlike>)
+
+#define REGISTER_RESULTTYPE(structName, rowsExpr, ColsExpr, MaxRowsExpr, MaxColsExpr)      \
+  REGISTER_RESULTTYPE_IMPL(structName, rowsExpr, ColsExpr, Ikarus::Impl::VectorizeGeneric, \
+                           Ikarus::Impl::MatricizeGeneric)
+
+#define REGISTER_RESERVED_RESULTTYPE(structName, rowsExpr, ColsExpr, MaxRowsExpr, MaxColsExpr)                       \
+  REGISTER_RESULTTYPE_IMPL(structName, rowsExpr, ColsExpr, MaxRowsExpr, MaxColsExpr, Ikarus::Impl::VectorizeGeneric, \
+                           Ikarus::Impl::MatricizeGeneric)
+
+#define REGISTER_SIMPLE_RESULTTYPE(structName, rowsExpr, ColsExpr) \
+  REGISTER_RESERVED_RESULTTYPE(structName, rowsExpr, ColsExpr, rowsExpr, ColsExpr)
+
+  ///=============================
+  /// Registered ResultTypes
+
+  REGISTER_SIMPLE_SYMMETRIC_RESULTTYPE(linearStress, worldDim, worldDim, false);
+  REGISTER_SIMPLE_SYMMETRIC_RESULTTYPE(PK2Stress, worldDim, worldDim, false);
+  REGISTER_SIMPLE_SYMMETRIC_RESULTTYPE(cauchyStress, worldDim, worldDim, false);
+
+  REGISTER_SIMPLE_RESULTTYPE(director, worldDim, 1);
+  REGISTER_SIMPLE_RESULTTYPE(magnetization, worldDim, 1);
+  REGISTER_SIMPLE_RESULTTYPE(gradientNormOfMagnetization, 1, 1);
+  REGISTER_SIMPLE_RESULTTYPE(vectorPotential, worldDim, 1);
+  REGISTER_SIMPLE_RESULTTYPE(divergenceOfVectorPotential, 1, 1);
+
+  REGISTER_SIMPLE_RESULTTYPE(BField, worldDim, 1);
+  REGISTER_SIMPLE_RESULTTYPE(HField, worldDim, 1);
+
+  REGISTER_SIMPLE_RESULTTYPE(customType, Eigen::Dynamic, Eigen::Dynamic);
+} // namespace ResultType
+
+template <template <typename, int, int> class RT1, template <typename, int, int> class RT2>
+constexpr bool isSameResultType = std::is_same_v<RT1<double, 1, 1>, RT2<double, 1, 1>>;
+
+/**
+ * \brief Container that is used for FE Results. It gives access to the stored value, but can also be used to access
+ * the result in Matrix and Vector form
+ * \tparam RT A specified ResultType
+ * \tparam storedValueIsVector boolean indicating whether the stored result is in its vector form, defaults to true
+ */
+template <typename RT, bool storedValueIsVector = true>
```

Comment:
instead of `bool` we can use an enum ResultShape with `Matrixshape`and vectorshape` or something better then below at call site, there is not this magic `true` inside calculateAt, which makes it more readable.